### PR TITLE
Provide an option to capture libvirt debug logs

### PR DESCRIPTION
**Pull request by anirudhrb**
Introduce a new option capture_libvirt_debug_logs in LibvirtPlatformSchema that tells LISA to capture libvirt debug logs during environment teardown.

Debug logs are enabled by using the log_outputs option in /etc/libvirt/libvirtd.conf. Here, we configure libvirt to write the debug logs to /var/log/libvirt/libvirtd.log.

We have been seeing multiple libvirt related errors in the pipeline and these logs can help in debugging those.

This is an opt-in feature since all users of libvirt platform might not be interested in debug level logs.

### CODE REVIEW

**Comment by squirrelsc** [comment]
@cwize1 please help on reviewing the change.

**Comment by cwize1** [comment]
What was the problem with the current journalctl solution?

**Comment by anirudhrb** [comment]
> What was the problem with the current journalctl solution?

No problem. That is still the default.

For debug logs libvirt recommends using a file as destination since journald might throttle the amount of logs.
https://libvirt.org/kbase/debuglogs.html

File: lisa/sut_orchestrator/libvirt/platform.py [hunk 1]
```diff
@@ -155,6 +167,11 @@ def _prepare_environment(self, environment: Environment, log: Logger) -> bool:
                 parent_logger=log,
             )
 
+        if self.platform_runbook.capture_libvirt_debug_logs:
+            self._enable_libvirt_debug_log()
+
+        self.libvirt_conn = libvirt.open(self.libvirt_conn_str)
```
Comment (cwize1) [inline lisa/sut_orchestrator/libvirt/platform.py:173]: Why was this line moved?

File: lisa/sut_orchestrator/libvirt/platform.py [hunk 2]
```diff
@@ -1187,3 +1199,29 @@ def _get_environment_information(self, environment: Environment) -> Dict[str, st
                     node.log.exception(f"error on get {key}.", exc_info=identifier)
 
         return information
+
+    def _enable_libvirt_debug_log(self) -> None:
```
Comment (cwize1) [inline lisa/sut_orchestrator/libvirt/platform.py:1216]: Should this be disabled after all the tests have run?

File: lisa/sut_orchestrator/libvirt/platform.py [hunk 3]
```diff
@@ -1187,3 +1199,29 @@ def _get_environment_information(self, environment: Environment) -> Dict[str, st
...
+        )
+
+        self.host_node.tools[Service].restart_service("libvirtd")
+
+    def _capture_libvirt_logs(self) -> None:
+        libvirt_log_path = self.host_node.local_log_path / "libvirtd.log"
+
+        if self.platform_runbook.capture_libvirt_debug_logs:
+            user = self.host_node.tools[Whoami].get_username()
+            self.host_node.tools[Chown].change_owner(self.LIBVIRT_DEBUG_LOG_PATH, user)
```
Comment (cwize1) [inline lisa/sut_orchestrator/libvirt/platform.py]: This seems like it might cause problems, especially if run on someone’s dev box. I recommend copying the file to `working_dir` first, changing the permissions on that file and then copying to the test runner machine.

File: lisa/sut_orchestrator/libvirt/platform.py (hunk 1 above)
Comment (cwize1) [reply lisa/sut_orchestrator/libvirt/platform.py:173]: If I recall correctly, I needed to place this line in `_initialize` to ensure that it is only opened once for all test suites/cases. Otherwise, the console logger would be unhappy when running multiple tests that include a reboot test.

**Review by cwize1 — CHANGES_REQUESTED** [review]
.

File: lisa/sut_orchestrator/libvirt/platform.py (hunk 1 above)
Comment (anirudhrb) [reply lisa/sut_orchestrator/libvirt/platform.py:173]: I moved it because _enable_libvirt_debug_log() restarts libvirtd. So, the connection is closed. Looks like I need to move this to stuff to `_initialize`.

File: lisa/sut_orchestrator/libvirt/platform.py (hunk 3 above)
Comment (anirudhrb) [reply lisa/sut_orchestrator/libvirt/platform.py]: What kind of problems do you expect this might cause?

File: lisa/sut_orchestrator/libvirt/platform.py (hunk 1 above)
Comment (anirudhrb) [reply lisa/sut_orchestrator/libvirt/platform.py:173]: Please check the latest iteration. I moved `host_node` also to `initialize` since it doesn't need to be done "per environment".

File: lisa/sut_orchestrator/libvirt/platform.py (hunk 3 above)
Comment (cwize1) [reply lisa/sut_orchestrator/libvirt/platform.py]: You shouldn’t change permissions in system directories to be that of a user.

**Review by cwize1 — CHANGES_REQUESTED** [review]
.

File: lisa/sut_orchestrator/libvirt/platform.py (hunk 3 above)
Comment (anirudhrb) [reply lisa/sut_orchestrator/libvirt/platform.py]: Done.

File: lisa/sut_orchestrator/libvirt/platform.py (hunk 2 above)
Comment (anirudhrb) [reply lisa/sut_orchestrator/libvirt/platform.py:1216]: Looks to me like there isn't a good place to do this. `_delete_environment` being called doesn't necessarily mean that all tests have run right? If an environment goes down during test execution lisa will create a new one.
Do we need more lifecycle callbacks?

Comment (squirrelsc) [reply lisa/sut_orchestrator/libvirt/platform.py:1216]: Good point, please add a "cleanup" method on platform level, and call it in lisa_runner.close.

Comment (anirudhrb) [reply lisa/sut_orchestrator/libvirt/platform.py:1216]: Done. Please take a look.

**Comment by anirudhrb** [comment]
@cwize1 @squirrelsc please do another round of review.

File: lisa/sut_orchestrator/libvirt/platform.py [hunk 4]
```diff
@@ -168,12 +189,10 @@ def _delete_environment(self, environment: Environment, log: Logger) -> None:
...
-        libvirt_log = self.host_node.tools[Journalctl].logs_for_unit(
-            "libvirtd", sudo=self.host_node.is_remote
-        )
-        libvirt_log_path = self.host_node.local_log_path / "libvirtd.log"
-        with open(str(libvirt_log_path), "w") as f:
-            f.write(libvirt_log)
+        self._capture_libvirt_logs()
+
+    def _cleanup(self) -> None:
+        self._disable_libvirt_debug_log()
```
Comment (cwize1) [inline lisa/sut_orchestrator/libvirt/platform.py]: Only run this if `capture_libvirt_debug_logs` is true?

**Review by cwize1 — DISMISSED** [review]

File: lisa/sut_orchestrator/libvirt/platform.py (hunk 4 above)
Comment (anirudhrb) [reply lisa/sut_orchestrator/libvirt/platform.py]: Done.

**Review by squirrelsc — APPROVED** [review]
